sql/tests: TestRandomSyntaxFunctions failed

sql/tests.TestRandomSyntaxFunctions [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3001942&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3001942&tab=artifacts#/) on master @ [bf71898e043692cc87d2c232fe3066054275ec29](https://github.com/cockroachdb/cockroach/commits/bf71898e043692cc87d2c232fe3066054275ec29):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0x33b252a]

```
Stack: 

```
goroutine 3702995 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).recordStatementSummary(0xc047417900, 0x5d57760, 0xc04737d470, 0xc047417e08, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor_statement_metrics.go:237 +0xb0a
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc047417900, 0x5d57760, 0xc04737d470, 0xc047417e08, 0x5db11e0, 0xc0476ff1a0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:934 +0x94b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc047417900, 0x5d57760, 0xc04737d470, 0x5d651a0, 0xc0405a6b90, 0x4ffe6fb, 0x63, 0x0, 0x1, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:667 +0x1006
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc047417900, 0x5d39160, 0xc01230d400, 0x5d651a0, 0xc0405a6b90, 0x4ffe6fb, 0x63, 0x0, 0x1, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:123 +0xb14
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0x5d651a0, 0xc0405a6b90, 0x4ffe6fb, 0x63, 0x0, 0x1, 0x156fa2c5, 0xed838a0d1, 0x0, 0x156fa2c5, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1580 +0x225
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc047417900, 0x5d39160, 0xc01230d400, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1582 +0x1457
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc047417900, 0x5d39220, 0xc0476feea0, 0xc00ae2a500, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1504 +0x228
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).initConnEx.func1(0xc047417900, 0x5d39220, 0xc0476feea0, 0xc01e08fc80, 0xc0476fef60, 0x5ce02e0, 0xc01230d240, 0xc042c57ef0, 0xc035e7ca00)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:203 +0x65
created by github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).initConnEx
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:202 +0x2f2
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRandomSyntaxFunctions
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRandomSyntaxFunctions672763105
    test_log_scope.go:74: use -show-logs to present logs inline
    rsg_test.go:731: 5s of 5m0s: 23576 executions, 14947 successful
    rsg_test.go:731: 10s of 5m0s: 46965 executions, 29799 successful
    rsg_test.go:731: 15s of 5m0s: 70860 executions, 45010 successful
    rsg_test.go:731: 20s of 5m0s: 94433 executions, 59926 successful
    rsg_test.go:731: 25s of 5m0s: 118488 executions, 75205 successful
    rsg_test.go:731: 30s of 5m0s: 141729 executions, 89978 successful
    rsg_test.go:731: 35s of 5m0s: 164833 executions, 104656 successful
    rsg_test.go:731: 40s of 5m0s: 187864 executions, 119321 successful
    rsg_test.go:731: 45s of 5m0s: 211470 executions, 134346 successful
    rsg_test.go:731: 50s of 5m0s: 233982 executions, 147977 successful
    rsg_test.go:731: 55s of 5m0s: 256772 executions, 160541 successful
```
</p>
</details>
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestRandomSyntaxFunctions PKG=./pkg/sql/tests TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #63943 sql/tests: TestRandomSyntaxFunctions failed [C-test-failure O-robot branch-release-20.2]
</p>
</details>
<details><summary>Internal log</summary>
<p>

```
mjibson marked as alumn{us/a}; resolving to rafiss instead
```
</p>
</details>

/cc @cockroachdb/sql-experience rafiss
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxFunctions.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/tests.TestRandomSyntaxFunctions [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3005160&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3005160&tab=artifacts#/) on master @ [d46b65000118d7f401d61f90a11d0a2438949216](https://github.com/cockroachdb/cockroach/commits/d46b65000118d7f401d61f90a11d0a2438949216):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0x33b406a]

```
Stack: 

```
goroutine 4439676 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).recordStatementSummary(0xc01cd1e600, 0x5d5bb00, 0xc01d970930, 0xc01cd1eb08, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor_statement_metrics.go:249 +0xb0a
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc01cd1e600, 0x5d5bb00, 0xc01d970930, 0xc01cd1eb08, 0x5db56a0, 0xc03b92b4d0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:934 +0x94b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc01cd1e600, 0x5d5bb00, 0xc01d970930, 0x5d69540, 0xc03db5e3c0, 0xc03905ccb0, 0x64, 0x2, 0x0, 0xc03f1a3e00, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:667 +0x1006
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc01cd1e600, 0x5d5bb00, 0xc01d970750, 0x5d69540, 0xc03db5e3c0, 0xc03905ccb0, 0x64, 0x2, 0x0, 0xc03f1a3e00, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:123 +0xb14
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execPortal(0xc01cd1e600, 0x5d5bb00, 0xc01d970750, 0xc03f1a3e00, 0xc00ce56000, 0x2, 0x2, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:210 +0x14e
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func2(0x0, 0x0, 0x0, 0x20eb2ac0, 0xed839f2c7, 0x0, 0xc01cd1e600, 0xc03aaedbe0, 0xc03aaedbd0, 0x2, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1648 +0x47b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc01cd1e600, 0x5d3d500, 0xc0065a4ec0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1650 +0x45c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc01cd1e600, 0x5d5bb00, 0xc01d5cbfb0, 0xc018b2b680, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1504 +0x228
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).initConnEx.func1(0xc01cd1e600, 0x5d5bb00, 0xc01d5cbfb0, 0xc014bee000, 0xc03b92b110, 0x5ce48c0, 0xc0065a4cc0, 0xc01d2e9710, 0xc01ff78750)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:203 +0x65
created by github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).initConnEx
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:202 +0x2f2
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRandomSyntaxFunctions
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRandomSyntaxFunctions282192907
    test_log_scope.go:74: use -show-logs to present logs inline
    rsg_test.go:731: 5s of 5m0s: 24727 executions, 15715 successful
    rsg_test.go:731: 10s of 5m0s: 48242 executions, 30635 successful
    rsg_test.go:731: 15s of 5m0s: 72263 executions, 45919 successful
    rsg_test.go:731: 20s of 5m0s: 96368 executions, 61227 successful
    rsg_test.go:731: 25s of 5m0s: 120564 executions, 76626 successful
    rsg_test.go:731: 30s of 5m0s: 144853 executions, 92063 successful
    rsg_test.go:731: 35s of 5m0s: 168301 executions, 107036 successful
    rsg_test.go:731: 40s of 5m0s: 191913 executions, 122039 successful
    rsg_test.go:731: 45s of 5m0s: 214277 executions, 136335 successful
    rsg_test.go:731: 50s of 5m0s: 237741 executions, 151260 successful
    rsg_test.go:731: 55s of 5m0s: 261455 executions, 166333 successful
    rsg_test.go:731: 1m0s of 5m0s: 284586 executions, 180982 successful
    rsg_test.go:731: 1m5s of 5m0s: 307807 executions, 195755 successful
    rsg_test.go:731: 1m10s of 5m0s: 330534 executions, 208794 successful
    rsg_test.go:731: 1m15s of 5m0s: 352768 executions, 220915 successful
    rsg_test.go:731: 1m20s of 5m0s: 375761 executions, 233603 successful
    rsg_test.go:731: 1m25s of 5m0s: 398223 executions, 245811 successful
    rsg_test.go:731: 1m30s of 5m0s: 420512 executions, 257941 successful
```
</p>
</details>
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestRandomSyntaxFunctions PKG=./pkg/sql/tests TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #63943 sql/tests: TestRandomSyntaxFunctions failed [C-test-failure O-robot branch-release-20.2]
</p>
</details>
<details><summary>Internal log</summary>
<p>

```
mjibson marked as alumn{us/a}; resolving to rafiss instead
```
</p>
</details>

/cc @cockroachdb/sql-experience rafiss
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxFunctions.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/conn_executor.go
+++ b/pkg/sql/conn_executor.go
@@ -324,28 +324,6 @@ func NewServer(cfg *ExecutorConfig, pool *mon.BytesMonitor) *Server {
 }
 
 func makeMetrics(cfg *ExecutorConfig, internal bool) Metrics {
-	var statsMetrics StatsMetrics
-	if !internal {
-		statsMetrics = StatsMetrics{
-			SQLStatsMemoryMaxBytesHist: metric.NewHistogram(
-				getMetricMeta(MetaSQLStatsMemMaxBytes, internal),
-				cfg.HistogramWindowInterval,
-				log10int64times1000,
-				3, /* sigFigs */
-			),
-			SQLStatsMemoryCurBytesCount: metric.NewGauge(
-				getMetricMeta(MetaSQLStatsMemCurBytes, internal)),
-			ReportedSQLStatsMemoryMaxBytesHist: metric.NewHistogram(
-				getMetricMeta(MetaReportedSQLStatsMemMaxBytes, internal),
-				cfg.HistogramWindowInterval,
-				log10int64times1000,
-				3, /* sigFigs */
-			),
-			ReportedSQLStatsMemoryCurBytesCount: metric.NewGauge(
-				getMetricMeta(MetaReportedSQLStatsMemCurBytes, internal)),
-			DiscardedStatsCount: metric.NewCounter(getMetricMeta(MetaDiscardedSQLStats, internal)),
-		}
-	}
 	return Metrics{
 		EngineMetrics: EngineMetrics{
 			DistSQLSelectCount:    metric.NewCounter(getMetricMeta(MetaDistSQLSelect, internal)),
@@ -371,7 +349,25 @@ func makeMetrics(cfg *ExecutorConfig, internal bool) Metrics {
 		},
 		StartedStatementCounters:  makeStartedStatementCounters(internal),
 		ExecutedStatementCounters: makeExecutedStatementCounters(internal),
-		StatsMetrics:              statsMetrics,
+		StatsMetrics: StatsMetrics{
+			SQLStatsMemoryMaxBytesHist: metric.NewHistogram(
+				getMetricMeta(MetaSQLStatsMemMaxBytes, internal),
+				cfg.HistogramWindowInterval,
+				log10int64times1000,
+				3, /* sigFigs */
+			),
+			SQLStatsMemoryCurBytesCount: metric.NewGauge(
+				getMetricMeta(MetaSQLStatsMemCurBytes, internal)),
+			ReportedSQLStatsMemoryMaxBytesHist: metric.NewHistogram(
+				getMetricMeta(MetaReportedSQLStatsMemMaxBytes, internal),
+				cfg.HistogramWindowInterval,
+				log10int64times1000,
+				3, /* sigFigs */
+			),
+			ReportedSQLStatsMemoryCurBytesCount: metric.NewGauge(
+				getMetricMeta(MetaReportedSQLStatsMemCurBytes, internal)),
+			DiscardedStatsCount: metric.NewCounter(getMetricMeta(MetaDiscardedSQLStats, internal)),
+		},
 	}
 }
 

--- a/pkg/sql/internal_test.go
+++ b/pkg/sql/internal_test.go
@@ -173,6 +173,33 @@ func TestInternalFullTableScan(t *testing.T) {
 	require.NoError(t, err)
 }
 
+// Test for regression https://github.com/cockroachdb/cockroach/issues/65523
+func TestInternalStmtFingerprintLimit(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	params, _ := tests.CreateTestServerParams()
+	s, db, _ := serverutils.StartServer(t, params)
+	defer s.Stopper().Stop(ctx)
+
+	_, err := db.Exec("SET CLUSTER SETTING sql.metrics.max_mem_app_txn_fingerprints = 0;")
+	require.NoError(t, err)
+
+	_, err = db.Exec("SET CLUSTER SETTING sql.metrics.max_mem_app_stmt_fingerprints = 0;")
+	require.NoError(t, err)
+
+	ie := sql.MakeInternalExecutor(
+		ctx,
+		s.(*server.TestServer).Server.PGServer().SQLServer,
+		sql.MemoryMetrics{},
+		s.ExecutorConfig().(sql.ExecutorConfig).Settings,
+	)
+
+	_, err = ie.Exec(ctx, "stmt-exceeds-fingerprint-limit", nil, "SELECT 1")
+	require.NoError(t, err)
+}
+
 func TestQueryIsAdminWithNoTxn(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)

--- a/pkg/sql/pgwire/server.go
+++ b/pkg/sql/pgwire/server.go
@@ -351,6 +351,7 @@ func (s *Server) Metrics() (res []interface{}) {
 		&s.SQLServer.InternalMetrics.StartedStatementCounters,
 		&s.SQLServer.InternalMetrics.ExecutedStatementCounters,
 		&s.SQLServer.InternalMetrics.EngineMetrics,
+		&s.SQLServer.InternalMetrics.StatsMetrics,
 	}
 }
 

--- a/pkg/ts/catalog/chart_catalog.go
+++ b/pkg/ts/catalog/chart_catalog.go
@@ -1714,6 +1714,26 @@ var charts = []sectionDescription{
 				Title:   "Number of fingerprint statistics being discarded",
 				Metrics: []string{"sql.stats.discarded.current"},
 			},
+			{
+				Title:   "Memory usage for internal fingerprint storage",
+				Metrics: []string{"sql.stats.mem.max.internal"},
+			},
+			{
+				Title:   "Current memory usage for internal fingerprint storage",
+				Metrics: []string{"sql.stats.mem.current.internal"},
+			},
+			{
+				Title:   "Memory usage for internal reported fingerprint storage",
+				Metrics: []string{"sql.stats.reported.mem.max.internal"},
+			},
+			{
+				Title:   "Current memory usage for internal reported fingerprint storage",
+				Metrics: []string{"sql.stats.reported.mem.current.internal"},
+			},
+			{
+				Title:   "Number of internal fingerprint statistics being discarded",
+				Metrics: []string{"sql.stats.discarded.current.internal"},
+			},
 		},
 	},
 	{

